core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailRepositoryConnection.java

```diff
@@ -155,7 +165,41 @@ public void begin() throws RepositoryException {
 	@Override
 	public void begin(IsolationLevel level) throws RepositoryException {
 		try {
-			sailConnection.begin(level);
+			// always call receiveTransactionSettings(...) before calling begin();
+			sailConnection.receiveTransactionSettings(new TransactionSetting[0]);
+
+			if (level != null) {
+				sailConnection.begin(level);
+			} else {
+				sailConnection.begin();
+			}
```

Comment:
This check is not necessary - you can just pass `null` as an argument to the `begin` method.

Reply:
I really prefer to move away from passing null to the begin method. We could consider deprecating and removing the `begin()` method and only keep the `begin(level)` method for the sail code.

Reply:
Also, from seeing that there were no test failures when I removed the null check from the SchemaCachingRDFSInferencerConnection I'm starting to suspect that we don't have any tests for this null handling behaviour.

Here is from the docs:

```
	/**
	 * Begins a transaction with the specified {@link IsolationLevel} level, requiring {@link #commit()} or
	 * {@link #rollback()} to be called to close the transaction.
	 *
	 * @param level the transaction isolation level on which this transaction operates.
	 * @throws UnknownSailTransactionStateException If the IsolationLevel is not supported by this implementation
	 * @throws SailException                        If the connection could not start a transaction, if the supplied
	 *                                              transaction isolation level is not supported, or if a transaction is
	 *                                              already active on this connection.
	 */
	void begin(IsolationLevel level) throws UnknownSailTransactionStateException, SailException;

```

From looking at that documentation from the SailConnection interface there is no mention of how null should be handled and it would make just as much sense to throw a SailException if null were provided.

Reply:
Mind you, the RepositoryConnection code specifically states that it allows null as an isolation level.

Reply:
> Also, from seeing that there were no test failures when I removed the null check from the SchemaCachingRDFSInferencerConnection I'm starting to suspect that we don't have any tests for this null handling behaviour.
> 
> Here is from the docs:
> 
> ```
> 	/**
> 	 * Begins a transaction with the specified {@link IsolationLevel} level, requiring {@link #commit()} or
> 	 * {@link #rollback()} to be called to close the transaction.
> 	 *
> 	 * @param level the transaction isolation level on which this transaction operates.
> 	 * @throws UnknownSailTransactionStateException If the IsolationLevel is not supported by this implementation
> 	 * @throws SailException                        If the connection could not start a transaction, if the supplied
> 	 *                                              transaction isolation level is not supported, or if a transaction is
> 	 *                                              already active on this connection.
> 	 */
> 	void begin(IsolationLevel level) throws UnknownSailTransactionStateException, SailException;
> ```
> 
> From looking at that documentation from the SailConnection interface there is no mention of how null should be handled and it would make just as much sense to throw a SailException if null were provided.

Spec-wise perhaps, but it wouldn't in practice: most (base) SailConnection implementations inherit from `AbstractSailConnection`, which accepts `null` and falls back to the default isolation level. But you're right that it's not officially documented. That's an oversight I think. 

We can discuss if that needs to change some other time but I would suggest for the purpose of this PR to stick to how most Sails currently behave. That doesn't mean you need to take out this if-else construction by the way: I only commented because I considered it unnecessary, but it doesn't harm to have it there either.

Reply:
Btw. Created a new issue: https://github.com/eclipse/rdf4j/issues/2376